File: dqops/src/main/java/com/dqops/metadata/comparisons/TableComparisonConfigurationSpec.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.dqops.metadata.comparisons;

import com.dqops.checks.CheckTimeScale;
import com.dqops.checks.CheckType;
import com.dqops.metadata.basespecs.AbstractSpec;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.id.HierarchyId;
import com.dqops.metadata.id.HierarchyNodeResultVisitor;
import com.dqops.utils.docs.SampleStringsRegistry;
import com.dqops.utils.docs.SampleValueFactory;
import com.dqops.utils.serialization.IgnoreEmptyYamlSerializer;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import lombok.EqualsAndHashCode;
import lombok.ToString;

import java.util.Objects;

/**
 * Identifies a data comparison configuration between a parent table (the compared table) and the target table from another data source (a reference table).
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
@ToString(callSuper = false)
public class TableComparisonConfigurationSpec extends AbstractSpec {
    private static final ChildHierarchyNodeFieldMapImpl<TableComparisonConfigurationSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(AbstractSpec.FIELDS) {
        {
            put("grouping_columns", o -> o.groupingColumns);
        }
    };

    @JsonPropertyDescription("The name of the connection in DQOp where the reference table (the source of truth) is configured. " +
                             "When the connection name is not provided, DQOps will find the reference table on the connection of the parent table.")
    private String referenceTableConnectionName;

    @JsonPropertyDescription("The name of the schema where the reference table is imported into DQOps. The reference table's metadata must be imported into DQOps.")
    private String referenceTableSchemaName;

    @JsonPropertyDescription("The name of the reference table that is imported into DQOps. The reference table's metadata must be imported into DQOps.")
    private String referenceTableName;

    @JsonPropertyDescription("Optional custom SQL filter expression that is added to the SQL query that retrieves the data from the compared table. " +
            "This expression must be a SQL expression that will be added to the WHERE clause when querying the compared table.")
    private String comparedTableFilter;

    @JsonPropertyDescription("Optional custom SQL filter expression that is added to the SQL query that retrieves the data from the reference table (the source of truth). " +
            "This expression must be a SQL expression that will be added to the WHERE clause when querying the reference table.")
    private String referenceTableFilter;

    @JsonPropertyDescription("The type of checks (profiling, monitoring, partitioned) that this check comparison configuration is applicable. The default value is 'profiling'.")
    private CheckType checkType = CheckType.profiling;

    @JsonPropertyDescription("The time scale that this check comparison configuration is applicable. Supported values are 'daily' and 'monthly' for monitoring and partitioned checks or an empty value for profiling checks.")
    private CheckTimeScale timeScale;

    @JsonPropertyDescription("List of column pairs from both the compared table and the reference table that are used in a GROUP BY clause  " +
            "for grouping both the compared table and the reference table (the source of truth). " +
            "The columns are used in the next of the table comparison to join the results of data groups (row counts, sums of columns) between the compared table and the reference table to compare the differences.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    @JsonSerialize(using = IgnoreEmptyYamlSerializer.class)
    private TableComparisonGroupingColumnsPairsListSpec groupingColumns = new TableComparisonGroupingColumnsPairsListSpec();


    /**
     * Returns the name of the connection (data source) in DQOps where the reference table is imported.
     * @return The name of the connection of the reference table.
     */
    public String getReferenceTableConnectionName() {
        return referenceTableConnectionName;
    }

    /**
     * Sets the name of the connection where the reference table is imported.
     * @param referenceTableConnectionName The name of the connection where the reference table is imported.
     */
    public void setReferenceTableConnectionName(String referenceTableConnectionName) {
        this.setDirtyIf(!Objects.equals(this.referenceTableConnectionName, referenceTableConnectionName));
        this.referenceTableConnectionName = referenceTableConnectionName;
    }

    /**
     * Returns the name of the schema where the reference table was imported.
     * @return The name of the schema with the reference table.
     */
    public String getReferenceTableSchemaName() {
        return referenceTableSchemaName;
    }

    /**
     * Sets the name of the schema with the reference table.
     * @param referenceTableSchemaName The name of the schema.
     */
    public void setReferenceTableSchemaName(String referenceTableSchemaName) {
        this.setDirtyIf(!Objects.equals(this.referenceTableSchemaName, referenceTableSchemaName));
        this.referenceTableSchemaName = referenceTableSchemaName;
    }

    /**
     * Returns the reference table name inside the schema.
     * @return The name of the reference table.
     */
    public String getReferenceTableName() {
        return referenceTableName;
    }

    /**
     * Sets the name of the reference table inside the schema.
     * @param referenceTableName Reference table name.
     */
    public void setReferenceTableName(String referenceTableName) {
        this.setDirtyIf(!Objects.equals(this.referenceTableName, referenceTableName));
        this.referenceTableName = referenceTableName;
    }

    /**
     * Returns the SQL expression that is added to the sensor query on the compared table.
     * @return SQL filter for the compared table.
     */
    public String getComparedTableFilter() {
        return comparedTableFilter;
    }

    /**
     * Sets the sql query that is applied on the compared table.
     * @param comparedTableFilter SQL filter applied on the compared table.
     */
    public void setComparedTableFilter(String comparedTableFilter) {
        this.setDirtyIf(!Objects.equals(this.comparedTableFilter, comparedTableFilter));
        this.comparedTableFilter = comparedTableFilter;
    }

    /**
     * Returns the SQL expression that is added to the sensor query on the reference table.
     * @return SQL filter for the reference table.
     */
    public String getReferenceTableFilter() {
        return referenceTableFilter;
    }

    /**
     * Sets the SQL query added to the query on the reference table.
     * @param referenceTableFilter SQL query for the reference table.
     */
    public void setReferenceTableFilter(String referenceTableFilter) {
        this.setDirtyIf(!Objects.equals(this.referenceTableFilter, referenceTableFilter));
        this.referenceTableFilter = referenceTableFilter;
    }

    /**
     * Returns the check type that this table comparison configuration is applicable to.
     * @return The type of checks where this comparison is used.
     */
    public CheckType getCheckType() {
        return checkType;
    }

    /**
     * Sets the type of checks where this comparison is used.
     * @param checkType Type of checks.
     */
    public void setCheckType(CheckType checkType) {
        this.setDirtyIf(!Objects.equals(this.checkType, checkType));
        this.checkType = checkType;
    }

    /**
     * Returns the time scale that this table comparison is applicable to.
     * @return Time scale.
     */
    public CheckTimeScale getTimeScale() {
        return timeScale;
    }

    /**
     * Sets the time scale of the checks (monitoring, partitioned) that are used.
     * @param timeScale Time scale.
     */
    public void setTimeScale(CheckTimeScale timeScale) {
        this.setDirtyIf(!Objects.equals(this.timeScale, timeScale));
        this.timeScale = timeScale;
    }

    /**
     * Returns the list of grouping column pairs used to group results before matching.
     * @return A list of column pairs used for grouping and matching.
     */
    public TableComparisonGroupingColumnsPairsListSpec getGroupingColumns() {
        return groupingColumns;
    }

    /**
     * Sets a list of columns used for grouping and matching.
     * @param groupingColumns A new list of column pairs.
     */
    public void setGroupingColumns(TableComparisonGroupingColumnsPairsListSpec groupingColumns) {
        setDirtyIf(!Objects.equals(this.groupingColumns, groupingColumns));
        this.groupingColumns = groupingColumns;
        propagateHierarchyIdToField(groupingColumns, "grouping_columns");
    }

    /**
     * Retrieves the comparison name from the parent hierarchy. Works only when the comparison specification was added to the comparison's map under a comparison name.
     * @return Comparison name.
     */
    @JsonIgnore
    public String getComparisonName() {
        HierarchyId hierarchyId = this.getHierarchyId();
        if (hierarchyId == null) {
            return null;
        }

        return hierarchyId.getLast().toString();
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Calls a visitor (using a visitor design pattern) that returns a result.
     *
     * @param visitor   Visitor instance.
     * @param parameter Additional parameter that will be passed back to the visitor.
     * @return Result value returned by an "accept" method of the visitor.
     */
    @Override
    public <P, R> R visit(HierarchyNodeResultVisitor<P, R> visitor, P parameter) {
        return visitor.accept(this, parameter);
    }

    /**
     * Creates and returns a copy of this object.
     */
    @Override
    public TableComparisonConfigurationSpec deepClone() {
        TableComparisonConfigurationSpec cloned = (TableComparisonConfigurationSpec) super.deepClone();
        return cloned;
    }

    public static class TableComparisonConfigurationSpecSampleFactory implements SampleValueFactory<TableComparisonConfigurationSpec> {
        @Override
        public TableComparisonConfigurationSpec createSample() {
            return new TableComparisonConfigurationSpec() {{
                setReferenceTableConnectionName(SampleStringsRegistry.getConnectionName());
                setReferenceTableSchemaName(SampleStringsRegistry.getSchemaName());
                setReferenceTableName(SampleStringsRegistry.getTableName());
            }};
        }
    }
}

File: distribution/python/dqops/client/models/table_comparison_configuration_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.check_time_scale import CheckTimeScale
from ..models.check_type import CheckType
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.table_comparison_grouping_columns_pair_spec import (
        TableComparisonGroupingColumnsPairSpec,
    )


T = TypeVar("T", bound="TableComparisonConfigurationSpec")


@_attrs_define
class TableComparisonConfigurationSpec:
    """
    Attributes:
        reference_table_connection_name (Union[Unset, str]): The name of the connection in DQOp where the reference
            table (the source of truth) is configured. When the connection name is not provided, DQOps will find the
            reference table on the connection of the parent table.
        reference_table_schema_name (Union[Unset, str]): The name of the schema where the reference table is imported
            into DQOps. The reference table's metadata must be imported into DQOps.
        reference_table_name (Union[Unset, str]): The name of the reference table that is imported into DQOps. The
            reference table's metadata must be imported into DQOps.
        compared_table_filter (Union[Unset, str]): Optional custom SQL filter expression that is added to the SQL query
            that retrieves the data from the compared table. This expression must be a SQL expression that will be added to
            the WHERE clause when querying the compared table.
        reference_table_filter (Union[Unset, str]): Optional custom SQL filter expression that is added to the SQL query
            that retrieves the data from the reference table (the source of truth). This expression must be a SQL expression
            that will be added to the WHERE clause when querying the reference table.
        check_type (Union[Unset, CheckType]):
        time_scale (Union[Unset, CheckTimeScale]):
        grouping_columns (Union[Unset, List['TableComparisonGroupingColumnsPairSpec']]): List of column pairs from both
            the compared table and the reference table that are used in a GROUP BY clause  for grouping both the compared
            table and the reference table (the source of truth). The columns are used in the next of the table comparison to
            join the results of data groups (row counts, sums of columns) between the compared table and the reference table
            to compare the differences.
    """

    reference_table_connection_name: Union[Unset, str] = UNSET
    reference_table_schema_name: Union[Unset, str] = UNSET
    reference_table_name: Union[Unset, str] = UNSET
    compared_table_filter: Union[Unset, str] = UNSET
    reference_table_filter: Union[Unset, str] = UNSET
    check_type: Union[Unset, CheckType] = UNSET
    time_scale: Union[Unset, CheckTimeScale] = UNSET
    grouping_columns: Union[Unset, List["TableComparisonGroupingColumnsPairSpec"]] = (
        UNSET
    )
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        reference_table_connection_name = self.reference_table_connection_name
        reference_table_schema_name = self.reference_table_schema_name
        reference_table_name = self.reference_table_name
        compared_table_filter = self.compared_table_filter
        reference_table_filter = self.reference_table_filter
        check_type: Union[Unset, str] = UNSET
        if not isinstance(self.check_type, Unset):
            check_type = self.check_type.value

        time_scale: Union[Unset, str] = UNSET
        if not isinstance(self.time_scale, Unset):
            time_scale = self.time_scale.value

        grouping_columns: Union[Unset, List[Dict[str, Any]]] = UNSET
        if not isinstance(self.grouping_columns, Unset):
            grouping_columns = []
            for grouping_columns_item_data in self.grouping_columns:
                grouping_columns_item = grouping_columns_item_data.to_dict()

                grouping_columns.append(grouping_columns_item)

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if reference_table_connection_name is not UNSET:
            field_dict["reference_table_connection_name"] = (
                reference_table_connection_name
            )
        if reference_table_schema_name is not UNSET:
            field_dict["reference_table_schema_name"] = reference_table_schema_name
        if reference_table_name is not UNSET:
            field_dict["reference_table_name"] = reference_table_name
        if compared_table_filter is not UNSET:
            field_dict["compared_table_filter"] = compared_table_filter
        if reference_table_filter is not UNSET:
            field_dict["reference_table_filter"] = reference_table_filter
        if check_type is not UNSET:
            field_dict["check_type"] = check_type
        if time_scale is not UNSET:
            field_dict["time_scale"] = time_scale
        if grouping_columns is not UNSET:
            field_dict["grouping_columns"] = grouping_columns

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.table_comparison_grouping_columns_pair_spec import (
            TableComparisonGroupingColumnsPairSpec,
        )

        d = src_dict.copy()
        reference_table_connection_name = d.pop(
            "reference_table_connection_name", UNSET
        )

        reference_table_schema_name = d.pop("reference_table_schema_name", UNSET)

        reference_table_name = d.pop("reference_table_name", UNSET)

        compared_table_filter = d.pop("compared_table_filter", UNSET)

        reference_table_filter = d.pop("reference_table_filter", UNSET)

        _check_type = d.pop("check_type", UNSET)
        check_type: Union[Unset, CheckType]
        if isinstance(_check_type, Unset):
            check_type = UNSET
        else:
            check_type = CheckType(_check_type)

        _time_scale = d.pop("time_scale", UNSET)
        time_scale: Union[Unset, CheckTimeScale]
        if isinstance(_time_scale, Unset):
            time_scale = UNSET
        else:
            time_scale = CheckTimeScale(_time_scale)

        grouping_columns = []
        _grouping_columns = d.pop("grouping_columns", UNSET)
        for grouping_columns_item_data in _grouping_columns or []:
            grouping_columns_item = TableComparisonGroupingColumnsPairSpec.from_dict(
                grouping_columns_item_data
            )

            grouping_columns.append(grouping_columns_item)

        table_comparison_configuration_spec = cls(
            reference_table_connection_name=reference_table_connection_name,
            reference_table_schema_name=reference_table_schema_name,
            reference_table_name=reference_table_name,
            compared_table_filter=compared_table_filter,
            reference_table_filter=reference_table_filter,
            check_type=check_type,
            time_scale=time_scale,
            grouping_columns=grouping_columns,
        )

        table_comparison_configuration_spec.additional_properties = d
        return table_comparison_configuration_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
